tests/accelerators/test_hpu.py

```diff
@@ -0,0 +1,242 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+
+import pytest
+import torch
+
+from pytorch_lightning import Callback, seed_everything, Trainer
+from pytorch_lightning.accelerators import HPUAccelerator
+from pytorch_lightning.strategies.hpu_parallel import HPUParallelStrategy
+from pytorch_lightning.strategies.single_hpu import SingleHPUStrategy
+from pytorch_lightning.utilities import _HPU_AVAILABLE
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from tests.helpers.boring_model import BoringModel
+from tests.helpers.datamodules import ClassifDataModule
+from tests.helpers.runif import RunIf
+from tests.helpers.simple_models import ClassificationModel
+
+
+@RunIf(hpu=True)
+def test_availability():
+    assert HPUAccelerator.is_available()
+
+
+@pytest.mark.skipif(_HPU_AVAILABLE, reason="test requires non-HPU machine")
+def test_fail_if_no_hpus():
+    with pytest.raises(MisconfigurationException, match="HPUAccelerator can not run on your system"):
+        Trainer(accelerator="hpu", devices=1)
+
+
+@RunIf(hpu=True)
+def test_accelerator_selected():
+    trainer = Trainer(accelerator="hpu")
+    assert isinstance(trainer.accelerator, HPUAccelerator)
+
+
+@RunIf(hpu=True)
+def test_all_stages(tmpdir, hpus):
+    """Tests all the model stages using BoringModel on HPU."""
+    model = BoringModel()
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        fast_dev_run=True,
+        accelerator="hpu",
+        devices=hpus,
+        precision=16,
+    )
+    trainer.fit(model)
+    trainer.validate(model)
+    trainer.test(model)
+    trainer.predict(model)
+
+
+@RunIf(hpu=True)
+def test_optimization(tmpdir):
+    seed_everything(42)
+
+    dm = ClassifDataModule(length=1024)
+    model = ClassificationModel()
+
+    trainer = Trainer(default_root_dir=tmpdir, max_epochs=1, accelerator="hpu", devices=1)
+
+    # fit model
+    trainer.fit(model, dm)
+    assert trainer.state.finished, f"Training failed with {trainer.state}"
+    assert dm.trainer is not None
+
+    # validate
+    result = trainer.validate(datamodule=dm)
+    assert dm.trainer is not None
+    assert result[0]["val_acc"] > 0.7
+
+    # test
+    result = trainer.test(model, datamodule=dm)
+    assert dm.trainer is not None
+    test_result = result[0]["test_acc"]
+    assert test_result > 0.6
+
+    # test saved model
+    model_path = os.path.join(tmpdir, "model.pt")
+    trainer.save_checkpoint(model_path)
+
+    model = ClassificationModel.load_from_checkpoint(model_path)
+
+    trainer = Trainer(default_root_dir=tmpdir, accelerator="hpu", devices=1)
+
+    result = trainer.test(model, datamodule=dm)
+    saved_result = result[0]["test_acc"]
+    assert saved_result == test_result
+
+
+@RunIf(hpu=True)
+def test_stages_correct(tmpdir):
+    """Ensure all stages correctly are traced correctly by asserting the output for each stage."""
```

Comment:
could you add more description of how the returning `dict` from each step works in the test as it is not common in other tests

Reply:
IMO, `TestCallback` callback explains it well on how the dict is returned after each step